### PR TITLE
fix(store): fix transaction rollback leaks, stale queries, undo diff corruption, and other store correctness bugs

### PR DESCRIPTION
**Pull request by steveruizok**
**Risk: high · Importance: high**

> **Umbrella PR — superseded by the per-bug splits below.** Kept open as the index and full write-up; close once the splits are in.
>
> History and listeners:
> - #10205 fix(store): drop rolled-back transaction changes from listeners and extractingChanges (supersedes #10100)
> - #10210 fix(store): don't deliver the in-flight flush to a listener added during it
> - #10216 fix(store): keep delivering a flush to other listeners when one throws
> - #10223 fix(store): flush pending changes to listeners on dispose
> - #10229 fix(store): attribute changes made by operationComplete handlers after a remote merge to the user
> - #10236 fix(store): stop store writes subscribing the writing reaction to the store
>
> Writes, diffs, and caches:
> - #10203 fix(store): record one history entry per record when put() sees the same id twice
> - #10207 fix(store): stop reverseRecordsDiff aliasing its input and squashRecordDiffs mutating shared tuples
> - #10215 fix(store): apply the removal of non-ephemeral keys in applyDiff with ignoreEphemeralKeys
> - #10225 fix(store): re-run computed cache readers when a deleted record comes back
> - #10231 fix(store): don't skip the next side-effect handler when one removes itself mid-dispatch
>
> Queries:
> - #10200 fix(store): rebuild indexes and queries after a rolled-back transaction resets the history
> - #10211 fix(store): make executeQuery and objectMatchesQuery agree on matcher semantics and empty expressions
> - #10218 fix(store): call the ids() query creator with no arguments
> - #10221 fix(store): don't emit empty added/removed sets from IncrementalSetConstructor
> - #10228 perf(store): avoid per-diff allocations in the index and ids derive loops
>
> Data structures, migrations, schema, docs:
> - #10202 fix(store): stop ImmutableMap overflowing the stack for keys named after Object.prototype members
> - #10209 fix(store): let AtomMap skip entries deleted mid-iteration and accept symbol and null-prototype keys
> - #10213 fix(store): don't report a duplicate migration dependency as circular
> - #10220 fix(store): return a migration error for malformed persisted schemas instead of throwing
> - #10227 fix(store): collect ids before deleting non-document records in migrateStorage
> - #10232 docs(store): use sentence case for DOCS.md headings

This PR fixes a batch of correctness bugs in `@tldraw/store` found by a deep review of the package: rolled-back transactions leaking to listeners and queries, undo recording states the document never had, reactions that write to the store subscribing to every store change, and a handful of edge cases in `AtomMap`, `ImmutableMap`, migrations, and schema handling. Every fix has a regression test that fails on `main`, and `packages/store/SPEC.md` is updated to match.

Supersedes #10100 (the accumulator rollback fix is the first item below, same approach). Relates to #10159.

### Before

Most of these are quiet: nothing throws, but sync or persistence receives a record the store never kept, a reactive query goes stale, or undo lands on a state that never existed.

**Transactions and history**

- `HistoryAccumulator` was not transaction-aware. `transact(() => { store.put([a]); throw })` rolled back `a` but left `{added: [a]}` queued, so the next flush delivered it to `TLSyncClient` and `TLLocalSyncClient`, which pushed or persisted a record the store did not hold.
- `StoreQueries.filterHistory` returned the raw history counter on `RESET`. After a rollback the counter can come back round to the value the incremental set last saw, so the reset registered no change and `ids()`/`index()` stayed permanently stale (phantom id present, next real id missing).
- `updateHistory` read `history.get()` with capture (and `AtomMap.deleteMany` captured the key set), so a reaction that writes to the store subscribed to every store change and could loop to the reaction depth limit.
- `put()` with the same id twice in one call recorded `updated[id] = [intermediate, final]`, or put the id in both `added` and `updated`. `editor.updateShapes([{id, x}, {id, y}])` therefore recorded a `from` the document never had, and undo restored it.
- `reverseRecordsDiff` aliased the input's `added`/`removed` containers, so `HistoryManager._undo` squashing into the reversed diff wrote through to the diff it had just pushed on the redo stack. `squashRecordDiffs([], {mutateFirstDiff: true})` also returned `undefined`, and `mutateFirstDiff` mutated shared tuples.
- `applyDiff({ignoreEphemeralKeys: true})` (the undo/redo path) merged only keys present in `to`, so a non-ephemeral key the update had removed was silently kept.
- `_flushHistory` iterated the live listener set (a listener added from a callback received the in-flight entries), `dispose()` dropped the pending change-set (an edit in the same frame as `Editor.dispose()` never reached sync/persistence), and a throwing listener starved every later listener.
- `flushAtomicCallbacks` only flipped `source` to `'user'` when after-handlers enqueued more events; events enqueued by `operationComplete` handlers after a remote merge ran as `'remote'`.

**Queries and caches**

- `executeQuery` (index walk) and `objectMatchesQuery` (predicate) disagreed on `{eq: undefined}`, `NaN`, multi-operator matchers, and empty nested sub-expressions, so the same `ids()` gave different answers from scratch and incrementally; `executeQuery({})` returned nothing, so `exec(type, {})` was `[]`.
- `createComputedCache` handed `UNINITIALIZED` to user `derive`/`areRecordsEqual` after the record was deleted, and a reader whose derive had yielded the same value before deletion never re-ran, so it never saw a re-created record with that id.
- Side-effect handler disposers spliced the live handler array during dispatch, so a handler that removed itself (a once-handler) made the next handler skip that event.

**Data structures, migrations, and schema**

- `ImmutableMap`'s string hash cache was a plain `{}`, so keys named after `Object.prototype` members (`'constructor'`, `'toString'`) got a function back as their hash; two such keys in a trie-sized map recursed until stack overflow. It also used `Object.is` instead of SameValueZero.
- `AtomMap` asserted instead of skipping when an entry was deleted mid-iteration (native `Map` skips), and `String(key)` threw for `Symbol` and null-prototype keys.
- `sortMigrations` double-counted the in-degree when `dependsOn` duplicated the implicit predecessor, asserting a false "Circular dependency in migrations".
- Malformed persisted schemas (missing `schemaVersion`/`sequences`, non-object schema, prototype-named sequence ids) threw `TypeError`s from `getMigrationsSince` instead of yielding an error result, bypassing `TLLocalSyncClient`'s error handling; legacy cleanup in `migrateStorage` deleted while iterating live storage.

### After

- Each accumulator entry is tagged with the history version it set; a rollback resets the counter, so later adds and flushes discard entries at or past it. `extractingChanges` excludes rolled-back nested transactions the same way. `filterHistory` returns a strictly increasing value on reset.
- `updateHistory` and `deleteMany` read without capture. `put()` keeps the first `from` per id, folds an update into a same-call addition, and records nothing for A → B → A.
- `reverseRecordsDiff` copies its collections; `squashRecordDiffs` claims tuples before mutating and returns a diff for empty input. `applyDiff` deletes keys absent from `to`.
- `_flushHistory` iterates a snapshot, delivers to every listener and then rethrows the first error; `dispose()` flushes before cancelling the reactor. `flushAtomicCallbacks` attributes every cascade round after the first to `'user'`.
- One shared `matchesValue` (SameValueZero; `undefined` never matches) backs both query paths, and an empty expression matches all records of the type.
- `createComputedCache` uses a never-equal `DELETED_RECORD` sentinel that cannot escape `get()`. Side-effect handler arrays are copy-on-write.
- `ImmutableMap` uses a null-prototype hash cache and SameValueZero; `AtomMap` iteration skips deleted entries and formats any key type. Duplicate migration edges are skipped. `getMigrationsSince` validates the persisted schema shape and returns error results; legacy cleanup collects ids before deleting.

### Implementation notes

- `Store.dispose()` now synchronously delivers pending entries to still-attached listeners. In `Editor.dispose()` the editor's own listeners and the sync/local clients are torn down first, so nothing new fires in practice, but a listener that outlives the editor now gets a final callback.
- `Store.createCache` (low-level) still exposes `UNINITIALIZED` to user signals after deletion; changing that is a type change, so it is documented rather than changed.
- Deliberately left for follow-ups: `put()` inside an enclosing transaction whose caller catches the error (interacts with the in-flight nested-transaction work in `packages/state`); interceptors still see rolled-back entries synchronously (SPEC H11, relied on by `HistoryManager`); the duplicated sections in `packages/store/DOCS.md`.

### Change type

- [x] `bugfix`

### Test plan

- [x] Unit tests — each new test fails on `main` and passes with this change (`packages/store` 498/498, plus `sync-core`, editor `HistoryManager`, and `tldraw` suites green)

### Release notes

- Fix rolled-back store transactions leaking changes to listeners, reactive queries, and `extractingChanges`
- Fix `put()` recording a wrong undo `from` when the same record appears twice in one call
- Fix `reverseRecordsDiff` aliasing its input and corrupting the redo stack
- Fix undo/redo keeping keys an update had removed
- Fix `executeQuery` and incremental query results disagreeing on `eq: undefined`, `NaN`, and empty expressions
- Fix computed caches never updating when a deleted record is re-created with the same id
- Fix a stack overflow in `AtomMap` for keys named after `Object.prototype` members
- Fix a false "Circular dependency in migrations" when `dependsOn` repeats the implicit predecessor
- Fix malformed persisted schemas throwing instead of returning a migration error

### API changes

- `QueryExpression` now carries its doc comment in the API report (a stray duplicate `@public` tag was hiding it); no signature changes

### Code changes

| Section                  | LOC change |
| ------------------------ | ---------- |
| Core code                | +362 / -178 |
| Tests                    | +675 / -3  |
| Automated files          | +1 / -1    |
| Documentation (SPEC/DOCS) | +77 / -70 |

### CODE REVIEW

File: packages/store/src/lib/executeQuery.ts [hunk 1]
```diff
@@ -147,6 +158,12 @@ export function executeQuery<R extends UnknownRecord, TypeName extends R['typeNa
 	// Extract all paths with matchers (flattens nested queries)
 	const matcherPaths = extractMatcherPaths(query)
 
+	// No matchers means no constraints, so every record of the type matches — as it does for
+	// `objectMatchesQuery`, for which an empty expression is vacuously true.
+	if (matcherPaths.length === 0) {
+		return store.getAllIdsForType(typeName)
+	}
```
[the remark this answers is not in the document]
Comment (steveruizok) [reply packages/store/src/lib/executeQuery.ts:175]: Good catch. Fixed in da19fb930e: `objectMatchesQuery` now skips a nested sub-expression only when it contains no matchers at any depth (new `hasAnyMatcher` helper), instead of checking `hasAnyKey`, so `{ metadata: { copies: {} } }` is match-all in both strategies like `extractMatcherPaths` already treats it. Added a QE3 agreement test for a nested no-matcher sub-expression on records that lack the nested object, and updated SPEC.md QE3 to state the rule.

**Comment by steveruizok** [comment]
Closing: superseded by the per-bug split indexed in #10253. The write-up above stays as the reference.
